docs/spec/consensus/checkvalidators.md

```diff
@@ -0,0 +1,224 @@
+\]
+
+
+
+
+Observe that validator 1 is not present in $nh.V$. Conservatively, we have to
+assume 1 is correct, and there may be a fault among 2,3,4. In addition, we don't
+know 5, so that conservatively, we have to assume 5 may be faulty. Thus among
+2,3,4,5, there may be two faults which violates the faults assumption. Thus $oh$
+does **not** provide sufficient trust in order to trust $nh$.
+
+
+## Basics for the "trusting period" method
+
+The function `CheckVS(oh, nh)` returns true, when *oh* provides sufficient
+trust to trust nh.
+
+### Assumptions
+
+1. $tp < unbonding period$.
+2. $nh.bfttime < now$
+3. $nh.bfttime < oh.bfttime+tp$
+4. $trust(oh)=true$
+
+### Some Maths
+
+**Observation 1.** If $oh.bfttime + tp > now$, we trust the old
+validator set $oh.V$.
+
+In the following let's assume oh is trusted and sufficiently new.
+
+**Definition 1.** Let $PA \subseteq oh.V$ be a *potential adversary* in $oh$, if
+the sum of the voting powers in PA is less than 1/3 of the voting
+powers in $oh.V$, that is,
+\[
+\sum_{(v,p) \in PA} p < 1/3 \sum_{(v,p) \in oh.V} p
+\]
+
+**Proposition 1.** The set of faulty processes
+\[oh.V \setminus \{(v,p): (v,p) \in oh.V \wedge correct(v,oh.bfttime + tp)\}\]
+is a potential adversary.
+
+*Proof.* By the LightClient invariant.
+
+**Definition 2.** Let the *unknown validators* UV be the validators that appear in $nh.V$ and not in
+$oh.V$, that is,
+\[
+UV = \{(v,p): (v,p) \in nh.V \wedge \nexists (v,x) \in oh.V \}.
+\]
+
+
+
+
+**Theorem 1.** If for all potential adversaries PA, in $nh$ the combined voting
+powers of PA and UV is less than a third of the total voting power, then in
+$nh$, more than 2/3 of the voting power is held by correct processes. Formally,
+if for all PA
+
+\[
+\sum_{(v,old) \in PA \wedge \\ (v,p) \in nh.V} p + \sum_{(v,p) \in UV} p < 1/3
+\sum_{(v,p) \in nh.V} p,
+\]
+then
+\[
+\sum_{(v,p) \in nh.V \wedge \\correct(v,oh.bfttime + tp)} p > 2/3 \sum_{(v,p) \in h.V} p
+\]
+
+*Proof.* By the definition of PA, Proposition 1, and the LightClient invariant.
+
+By Assumption 3, there is sufficient voting power to trust the new validator set. (And thus the validator set it signs in that block, for which the $tp$ starts at the $bfttime$ of the header).
+
+Below, we thus sketch a function that checks whether the premise of Theorem 1 holds. If the results is positive, we can trust $nh$, otherwise not.
+
+
+
+### An Algorithm
+
+In pseudo go...
+
+```go
+func CheckVS(oh, nh) bool {
+  if oh.bfttime + unbonding_period < now { // Observation 1
+    return false // old header was once trusted but it is expired
+  }
+
+  PAs := compute_all_PAs(oh)  // Definition 1
+  PAs := reduce (PAs)         // remove every PA that is a subset of another PA
+  UV := compute_UV(oh,nh)     // Definition 2
+
+  vpUV := votingpower(UV,nh)  // second sum in Theorem 1
+  vpNH := votingpower(nh.V,nh) // right hand side of premise of Theorem 1
+  vpMaxPA := maximumvotingpower(PAs,nh) // voting powers of all PA and big max
```

Comment:
Not sure if you looked into the complexity here yet. Maybe we can use the 0/1 knapsack algorithm (pseudo-polynomial), i.e. determine vpMaxPA, the maximum power that can fill a knapsack no bigger than 1/3 total voting power.
It would be O(n*vpNH), better than the O(2**n) for compute_all_PAs(oh) but still not great. 
n - number validators (~100)
vpNH - 1/3 total voting power, would have to look into real numbers but I think it is high...maybe can be optimized using normalization of powers (weights in knapsack problem)?.

Reply:
Haven't been thinking about efficient way of implementing it. We would need your help there :)

Reply:
See below comment.